TST: Fixed elements being shuffled

madphysicist opened the pull request:
Did not affect the current test results, but it caused a problem in some scipy code I was working on. Also, made a couple of doc changes.

The issue with only shuffling the first axis is that many of the combinations of dimensions will partition the relevance portions of the array right back into original order, guaranteeing that the answer will appear to be correct even if the indices were off.

Since the indices in all the affected tests were correct, this is more of a potential problem than an actual one, and even that with extremely low probability of ever manifesting.

madphysicist committed:
TST: Fixed shuffle axis in tests.

Since shuffle only works along the first dimension, it must be done before
reshape to get reasonable looking data. Did not affect the current tests. I
noticed while working on some scipy code.

Also, made a couple of doc changes to np.random.shuffle.

madphysicist commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4997,6 +4997,9 @@ cdef class RandomState:
 
         Modify a sequence in-place by shuffling its contents.
 
+        This function only shuffles the array along the first index of a
```
I moved this to the main function description because it is very important and I did not notice it in the example the fist time I read the docs.

rkern replied:
While we're here: `s/index/axis`.

rkern commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4997,6 +4997,9 @@ cdef class RandomState:
         Modify a sequence in-place by shuffling its contents.
 
+        This function only shuffles the array along the first index of a
+        multi-dimensional array:
```
No colon, if it's moved up here.

madphysicist replied:
Updated. Added a few words. Hope that's OK.

rkern commented:
One last comment, then LGTM.

madphysicist commented:
Fixed that and wrapped to 72 chars.

madphysicist commented:
I was testing `[ci skip]` with the second commit. I can squash them if desired.

charris commented:
Hmm, looks like the `[ci skip]` disabled the tests completely. @madphysicist Could you remove that so the tests run?

madphysicist commented:
Squashed, rebased and fixed message. The tests for the first commit out of the two were accessible via the "Commits" tab, which is why I did not squash before.

juliantaylor commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -2389,8 +2389,11 @@ def test_extended_axis(self):
+        # Shuffle only works along the first dimension:
+        # Reshape **after** shuffling.
         np.random.shuffle(d)
+        d = d.reshape((3, 5, 7, 11))
```
shuffle(d.ravel()) should also work but this is fine too

juliantaylor commented:
I wouldn't classify this as a bug, just sub-optimal test coverage as the function does still work with better tests
how can this have caused problems in scipy?

madphysicist commented:
@juliantaylor I have renamed from BUG to TST and made the commit message clearer. I also adopted the `shuffle(a.ravel())` notation you suggested since it looks much cleaner, even though it involves a tiny bit more actual work (reshape then unshape to shuffle).

This did not cause issues directly in scipy. I just noticed it because I copied over some of these tests to introduce an IQR function into scipy and the tests broke for one of the old versions of numpy. This was not all of the issue, but it was part of it.

juliantaylor commented:
ok thanks!
